ssl/ssl_sess.c

```diff
@@ -832,31 +885,59 @@ int SSL_SESSION_set1_id(SSL_SESSION *s, const unsigned char *sid,
 
 long SSL_SESSION_set_timeout(SSL_SESSION *s, long t)
 {
-    if (s == NULL)
+    time_t new_timeout = (time_t)t;
+
+    if (s == NULL || t < 0)
         return 0;
-    s->timeout = t;
+    if (s->timeout != new_timeout) {
+        if (s->owner != NULL) {
+            if (!CRYPTO_THREAD_write_lock(s->owner->lock))
+                return 0;
+            s->timeout = new_timeout;
+            ssl_session_calculate_timeout(s);
+            SSL_SESSION_list_add(s->owner, s);
+            CRYPTO_THREAD_unlock(s->owner->lock);
+        } else {
+            s->timeout = new_timeout;
+            ssl_session_calculate_timeout(s);
+        }
+    }
     return 1;
 }
 
 long SSL_SESSION_get_timeout(const SSL_SESSION *s)
 {
     if (s == NULL)
         return 0;
-    return s->timeout;
+    return (long)s->timeout;
 }
 
 long SSL_SESSION_get_time(const SSL_SESSION *s)
 {
     if (s == NULL)
         return 0;
-    return s->time;
+    return (long)s->time;
 }
 
 long SSL_SESSION_set_time(SSL_SESSION *s, long t)
 {
+    time_t new_time = (time_t)t;
+
     if (s == NULL)
         return 0;
-    s->time = t;
+    if (s->time != new_time) {
```

Comment:
This reads `s->time` with no lock, and then subsequently writes to it under a write lock. Shouldn't there be a read lock here?

Reply:
Not necessarily. It's checking to see if it's different; the actual value doesn't matter. This is an optimization to see if it should be changed. Previously it would just slam the time into the session without any locking at all. Any manipulation is done under lock. But note that the lock only happens if the session is in a SSL_CTX cache; the SSL_SESSION structure does not have a lock.

Reply:
But if `s->owner` is non-NULL then we know that the session has already been placed in the cache and therefore taking a read lock might be appropriate. It definitely seems odd to control writes with a write lock, but not reads.

Reply:
I would agree 100% if the value read were used for something meaningful. In this case, it's really not. It's read to see if there needs to be a change. If so, do the update; the prior value doesn't matter.
Because of that, locking for read here is pointless; if we're always going to lock, then we might as well always update, and remove the (albeit minor) optimization. I can make that change (unless my argument is convincing 😄 ).

Reply:
I would prefer to see it modified as you suggest.